File: Algorithms/0816_Ambiguous_Coordinates/Java/AmbiguousCoordinatesSolution1.java

// Space: O(1)
// Time: O(n)

import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Solution {
    public List<String> ambiguousCoordinates(String s) {
        String data = s.substring(1, s.length() - 1);
        List<String> res = new ArrayList<>();
        for (int i = 1; i < data.length(); i++) {
            List<String> leftRes = generateNumber(data.substring(0, i));
            List<String> rightRes = generateNumber(data.substring(i, data.length()));
            if (leftRes.size() == 0 || rightRes.size() == 0) continue;
            for (String left : leftRes) {
                for (String right : rightRes) {
                    String temp = "(" + left + ", " + right + ")";
                    res.add(temp);
                }
            }
        }
        return res;
    }

    public List<String> generateNumber(String s) {
        if (s.length() == 1) return new ArrayList<>(Collections.singletonList(s));
        List<String> res = new ArrayList<>();
        Pattern pattern = Pattern.compile("(^[1-9]+[0-9]*\\.[0-9]*[1-9]$)|(^0\\.[0-9]*[1-9]$)");

        int j = 0;
        while (j < s.length() && s.charAt(j) == '0') {
            j++;
        }
        String ltrim = s.substring(j);
        if (ltrim == s) res.add(s);

        for (int i = 1; i < s.length(); i++) {
            String temp = s.substring(0, i) + "." + s.substring(i, s.length());
            Matcher matcher = pattern.matcher(temp);
            if (matcher.find()) res.add(temp);
        }
        return res;
    }
}

File: Algorithms/0816_Ambiguous_Coordinates/Python/Ambiguous_Coordinates_Solution_1.py

# Space: O(1)
# Time: O(n)

import re

class Solution:
    def ambiguousCoordinates(self, s):
        data = s[1: -1]
        res = []

        for i in range(1, len(data)):
            left_res = self.generate_numbers(data[:i])
            right_res = self.generate_numbers(data[i:])

            if len(left_res) == 0 or len(right_res) == 0: continue

            for i in left_res:
                for j in right_res:
                    temp = '(' + i + ', ' + j + ')'
                    res.append(temp)

        return res

    def generate_numbers(self, string):
        regex_test = re.compile('(^[1-9]+[0-9]*\.[0-9]*[1-9]$)|(^0\.[0-9]*[1-9]$)')
        if len(string) == 1: return [string]
        res = []
        if string.lstrip('0') == string: res.append(string)

        for i in range(1, len(string)):
            temp = string[:i] + '.' + string[i:]
            if regex_test.match(temp): res.append(temp)

        return res





